pipenv/utils/resolver.py

```diff
@@ -189,12 +188,14 @@ def get_metadata(
         pre: bool = False,
         clear: bool = False,
     ) -> Tuple[
-        Set[str],
+        List[str],
         Dict[str, Dict[str, Union[str, bool, List[str]]]],
         Dict[str, str],
         Dict[str, str],
     ]:
-        constraints: Set[str] = set()
+        constraints: Dict[
+            str, None
+        ] = {}  # Used Dict instead of Set because Dict is ordered and is hence stable
```

Comment:
Why not a [frozenset](https://docs.python.org/3/library/stdtypes.html#frozenset)?